docs/dev/content/implementation.rst

```diff
@@ -0,0 +1,116 @@
+
+For example, this is how the client side dynamically request data from different channels:
+
+    >>> localhost:8000/api/content/<channel_1_id>/contentnode
+
+this will response with all the contentnode data stored in database channel_1_id.sqlite3
+
+    >>> localhost:8000/api/content/<channel_2_id>/contentnode
+
+this will response with all the contentnode data stored in database channel_2_id.sqlite3
+
+get_active_content_database
+---------------------------
+
+A utility function to retrieve the temporary thread-local variable that `using_content_database` sets
+
+set_active_content_database
+---------------------------
+
+A utility function to set the temporary thread-local variable
+
+using_content_database
+----------------------
+
+A decorator and context manager to do queries on a specific content DB.
+
+Usage as a context manager:
+
+    .. code-block:: python
+
+        from models import ContentNode
+
+        with using_content_database("nalanda"):
+            objects = ContentNode.objects.all()
+            return objects.count()
+
+Usage as a decorator:
+
+    .. code-block:: python
+
+        from models import ContentNode
+
+        @using_content_database('nalanda')
+        def delete_all_the_nalanda_content():
+            ContentNode.objects.all().delete()
+
+ContentDBRouter
+---------------
+
+A router that decides what content database to read from based on a thread-local variable.
+
+ContentNode
+-----------
+
+``ContentNode`` is implemented as a Django model that inherits from two abstract classes, MPTTModel and ContentDatabaseModel. 
+`django-mptt's MPTTModel <http://django-mptt.github.io/django-mptt/>`_, which
+allows for efficient traversal and querying of the ContentNode tree.
+``ContentDatabaseModel`` is used as a marker so that the content_db_router knows to query against the content database only if the model inherits the ContentDatabaseModel.
+
+The tree structure is established by the `parent` field that is a foreign key pointing to another ContentNode ojbect. You can also create a symmetric relationship using the `related` field, or an asymmetric field using the `is_prerequisite` field.
+
+File
+----
+
+The ``File`` model also inherits from ``ContentDatabaseModel``.
+
+To find where the source file is located, the class method ``get_url`` uses the ``checksum`` filed and ``settings.CONTENT_SOURCE_DIR`` to calculate the file path. Every source file is named based on its MD5 hash value(this value is also stored in the ``checksum`` field) and stored in namespaced folder under the directory specified in ``settings.CONTENT_SOURCE_DIR``. Because it's likely to have thousands of source files, some file system cannot handle flat folder with large number of files very well, we create namespaced subfolder to imporve the performance. So the eventual file path can look something like:
```

Comment:
"filed" -> "field"
"value(this" -> "value (this" (add space)
"in namespaced folder" -> "in a namespaced folder"
"thousands of source files, some" -> "thousands of source files, and some"
"file system" -> "filesystems"
"flat folder" -> "a flat folder"
"large number" -> "a large number"
"subfolder" -> "subfolders"
"imporve" -> "improve"
"can look" -> "would look"